File: internal/scriptruntime/protocol.go
package scriptruntime

import "github.com/brynnb/new-yokosuka-server/internal/scriptcontent"

const ProtocolVersion = "new-yokosuka-yarn-runtime-v1"

type Value struct {
	Type  string `json:"type"`
	Value string `json:"value"`
}

type Variable struct {
	Name  string `json:"name"`
	Value Value  `json:"value"`
}

type StartRequest struct {
	Program   []byte
	StartNode string
	Variables []Variable
}

type startMessage struct {
	ProtocolVersion string               `json:"protocolVersion"`
	ProgramBase64   string               `json:"programBase64"`
	StartNode       string               `json:"startNode"`
	Functions       []functionDefinition `json:"functions"`
	Variables       []Variable           `json:"variables"`
}

type functionDefinition struct {
	Name           string   `json:"name"`
	ReturnType     string   `json:"returnType"`
	ParameterTypes []string `json:"parameterTypes"`
}

type Input struct {
	Type     string `json:"type"`
	QueryID  *int   `json:"queryId,omitempty"`
	OptionID *int   `json:"optionId,omitempty"`
	Value    *Value `json:"value,omitempty"`
}

type Event struct {
	Type          string                           `json:"type"`
	Sequence      int                              `json:"sequence"`
	QueryID       *int                             `json:"queryId,omitempty"`
	Name          string                           `json:"name,omitempty"`
	Node          string                           `json:"node,omitempty"`
	LineID        string                           `json:"lineId,omitempty"`
	Substitutions []string                         `json:"substitutions,omitempty"`
	Arguments     []scriptcontent.CompiledArgument `json:"arguments,omitempty"`
	Options       []Option                         `json:"options,omitempty"`
	Message       string                           `json:"message,omitempty"`
}

type Option struct {
	ID            int      `json:"id"`
	LineID        string   `json:"lineId"`
	Substitutions []string `json:"substitutions"`
	IsAvailable   bool     `json:"isAvailable"`
}

func Continue() Input { return Input{Type: "continue"} }

func Cancel() Input { return Input{Type: "cancel"} }

func Select(optionID int) Input { return Input{Type: "select", OptionID: &optionID} }

func QueryResult(queryID int, value Value) Input {
	return Input{Type: "queryResult", QueryID: &queryID, Value: &value}
}

File: tools/yarn-compiler/Protocol.cs
namespace NewYokosuka.YarnCompiler;

internal sealed record CompileRequest(
    string FileName,
    string Source,
    IReadOnlyList<FunctionDefinition>? Functions
);

internal sealed record FunctionDefinition(
    string Name,
    string ReturnType,
    IReadOnlyList<string>? ParameterTypes
);

internal sealed record CompileResponse(
    string ProtocolVersion,
    string CompilerVersion,
    bool Valid,
    string? ProgramBase64,
    IReadOnlyList<CompilerDiagnostic> Diagnostics,
    IReadOnlyList<CompiledLine> Lines,
    IReadOnlyList<CompiledNode> Nodes,
    IReadOnlyList<CompiledCall> Calls
);

internal sealed record CompilerDiagnostic(
    string Code,
    string Severity,
    string Message,
    string FileName,
    int StartLine,
    int StartColumn,
    int EndLine,
    int EndColumn
);

internal sealed record CompiledLine(
    string Id,
    string? Text,
    string FileName,
    string NodeName,
    int LineNumber,
    bool HasImplicitId,
    IReadOnlyList<string> Metadata,
    string? ShadowLineId
);

internal sealed record CompiledNode(
    string Title,
    string? SourceTitle,
    string? UniqueTitle,
    string? Group,
    IReadOnlyList<string> FunctionCalls,
    IReadOnlyList<string> CommandCalls,
    IReadOnlyList<string> VariableReferences,
    IReadOnlyList<string> CharacterNames,
    IReadOnlyList<string> Tags,
    int OptionCount,
    int HeaderStartLine,
    int TitleLine,
    int BodyStartLine,
    int BodyEndLine
);

internal sealed record CompiledCall(
    string Kind,
    string Name,
    string Node,
    string FileName,
    int StartLine,
    int StartColumn,
    int EndLine,
    int EndColumn,
    string? ParseError,
    IReadOnlyList<CompiledArgument> Arguments
);

internal sealed record CompiledArgument(
    string Type,
    bool IsStatic,
    string? Value
);

internal sealed record RuntimeStartRequest(
    string ProtocolVersion,
    string ProgramBase64,
    string StartNode,
    IReadOnlyList<FunctionDefinition>? Functions,
    IReadOnlyList<RuntimeVariable>? Variables
);

internal sealed record RuntimeVariable(string Name, RuntimeValue Value);

internal sealed record RuntimeValue(string Type, string Value);

internal sealed record RuntimeInput(
    string Type,
    int? QueryId,
    int? OptionId,
    RuntimeValue? Value
);

internal sealed record RuntimeEvent(
    string Type,
    int Sequence,
    int? QueryId = null,
    string? Name = null,
    string? Node = null,
    string? LineId = null,
    IReadOnlyList<string>? Substitutions = null,
    IReadOnlyList<CompiledArgument>? Arguments = null,
    IReadOnlyList<RuntimeOption>? Options = null,
    string? Message = null
);

internal sealed record RuntimeOption(
    int Id,
    string LineId,
    IReadOnlyList<string> Substitutions,
    bool IsAvailable
);
